File: 2022-04/typescript/394_Decode_String.ts
function decodeString(s: string): string {
    const stack: string[] = []
    let numStr = ''
    let result = ''

    for (let c of s) {
        if (!isNaN(Number(c))) {
            numStr += c
        } else if (c === '[') {
            stack.push(result)
            stack.push(numStr)
            numStr = ''
            result = ''
        } else if (c === ']') {
            const n = Number(stack.pop())
            const prev = stack.pop()
            result = prev + (new Array(n).fill(result).join(''))
        } else {
            result += c
        }
    }
    return result
}

File: 2022-04/python/394_Decode_String.py
class Solution:
    def decodeString(self, s):
        stack = []
        num_str = ''
        result = ''
        for c in s:
            if c.isnumeric():
                num_str += c
            elif c == '[':
                stack.append(result)
                stack.append(num_str)
                num_str = ''
                result = ''
            elif c == ']':
                n = int(stack.pop())
                prev = stack.pop()
                result = prev + (result * n)
            else:
                result += c
        return result

